keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+      `tcpSocket:` to `tcp:` in their pod's probe definition.
+
+[`TCPSocketAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2248
+[`HTTPGetAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2219
+[Pod Security admission]: https://kubernetes.io/docs/concepts/security/pod-security-admission/
+[Happy Eyeballs]: https://tools.ietf.org/html/rfc8305
+
+### Non-Goals
+
+- Adding entirely-new probe types.
+
+- Changing the definition of NetworkPolicy v1 to remove the "kubelet
+  can access all local pods" hole, even in the case where the cluster
+  is configured in a way to forbid the use of probes that require the
+  hole. Even if _kubelet_ no longer needs to access pods, some users
+  likely depend on that rule to allow access to some pods from other
+  host-level processes, and there is no other way in NetworkPolicy v1
+  to write such a rule if we remove the built-in one. (However,
+  AdminNetworkPolicy and any future successor to NetworkPolicy should
+  be designed to not have this rule by default.)
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
+`TCPSocketAction.Host` and `HTTPGetAction.Host` fields, and allow them
+to be blocked by Pod Security admission.
+
+Beyond that, there are two (and a half) possible plans:
+
+1. Change the semantics of the existing `TCPSocketAction`,
+   `HTTPGetAction`, and `GRPCAction` probe types and say that
+   henceforth the probes will be done in the pod network namespace
+   rather than being sent over the pod network (unless you use the
+   deprecated `Host` field to request an alternate probe endpoint; we
+   probably shouldn't just move those kinds of probes into the pod
+   network namespace since that might require a "hairpin" connection
+   to work when it wasn't required before).
+
+2. Leave the existing probe types with their existing semantics, but
+   deprecate them and add new probe types with the
+   inside-the-network-namespace semantics. (`v1.ProbeHandler`
+   currently has `httpGet` and `tcpSocket` fields; these could
+   potentially become just `http` and `tcp`. It's less obvious what
+   `grpc` could be renamed to.)
+
+2½. Deprecate-and-replace `httpGet` and `tcpSocket` as in (2), but
+    change the semantics of `grpc` as in (1), since GRPC probes are
+    already more constrained than the TCP and HTTP probes anyway (and
+    I don't have a good suggestion for renaming them).
+
+### Kubelet/Runtime Changes to Fix the NetworkPolicy Hole
```

Comment:
@mrunalp : Could you PTAL this section and let us know your thoughts? (I think you and @aojea had talked about this at some point in the past)